tests/formatters.test.js

```diff
@@ -0,0 +1,85 @@
+import * as formatters from '../src/lib/formatters';
+
+test('serializeFilters method should return the updated search params', () => {
+  const testUrl = new URL('http://test.html?timeStamp=null&id=["dummyId123"]');
+  const searchParams = new URLSearchParams(testUrl.search.slice(1));
+
+  const extIds = ['addon1@test.com', 'addon2@test.com'];
+  const viewTypes = ['popup', undefined];
+  const types = ['api_call', 'content_script'];
+  const names = ['test@api'];
+  const tabId = 22;
+  const keyword = 'test@substring';
+  const timeStamp = { start: 1597686226302, stop: 1597686226302 };
+
+  const updateFilter = {
+    id: { exclude: new Set(extIds) },
+    viewType: { exclude: new Set(viewTypes) },
+    type: { exclude: new Set(types) },
+    name: { exclude: new Set(names) },
+    tabId,
+    keyword,
+    timeStamp,
+  };
+
+  const serializedFormat = formatters.serializeFilters(
+    searchParams,
+    updateFilter
+  );
+
+  const expectedExtIds = '["addon1@test.com","addon2@test.com"]';
+  const expectedViewTypes = '["popup",null]';
+  const expectedTypes = '["api_call","content_script"]';
+  const expectedNames = '["test@api"]';
+  const expectedTabId = '22';
+
+  expect(serializedFormat.get('id')).toEqual(expectedExtIds);
+  expect(serializedFormat.get('viewType')).toEqual(expectedViewTypes);
+  expect(serializedFormat.get('type')).toEqual(expectedTypes);
+  expect(serializedFormat.get('name')).toEqual(expectedNames);
+  expect(serializedFormat.get('tabId')).toEqual(expectedTabId);
+  expect(serializedFormat.get('keyword')).toEqual(keyword);
+  expect(serializedFormat.get('timeStamp')).toEqual(JSON.stringify(timeStamp));
+});
+
+test('deSerializeFilters method should return valid filter object', () => {
+  const testUrl = new URL(
+    'http://test.html?&id=["test1@addon","test2@addon"]&viewType=[null,"popup"]&type=["content_script","api_event"]&name=["tabs.executeScript"]&timeStamp={"start":123123,"stop":456456}&keyword=undefined&tabId=982'
+  );
+
+  const expectedFilterObject = {
+    id: { exclude: new Set(['test1@addon', 'test2@addon']) },
+    viewType: { exclude: new Set([undefined, 'popup']) },
+    type: { exclude: new Set(['content_script', 'api_event']) },
+    name: { exclude: new Set(['tabs.executeScript']) },
+    tabId: 982,
+    keyword: 'undefined',
+    timeStamp: { start: 123123, stop: 456456 },
+  };
+
+  const searchParams = new URLSearchParams(testUrl.search.slice(1));
+  const filterObject = formatters.deSerializeFilters(searchParams);
+
+  expect(filterObject).toMatchObject(expectedFilterObject);
```

Comment:
I'm not sure, can you verify that .toMatchObject actually checks the contents of the Set? That is, if you intentionally make the expectation invalid, does the test fail as expected?

Reply:
Yes, it checks the content of the Set too. :-)